rfcs/RFC-Uploads.md

```diff
@@ -0,0 +1,350 @@
+
+#### POST /dp/v1/dataset/{dataset_uuid}/asset/{asset_uuid}/upload
+
+Used to request an access token to continue uploading a file to S3 by an authenticated user who owns the data set. An
+error is returned if the file upload has been completed already.
+
+##### Request: POST /dp/v1/dataset/{dataset_uuid}/asset/{asset_uuid}/upload
+
+- data set_id - identifies the data set.
+- asset_id - identifies the asset.
+
+##### Response: POST /dp/v1/dataset/{dataset_uuid}/asset/{asset_uuid}/upload
+
+- Access Token - allows the user to upload their file to S3 in multiple parts.
+
+### AWS STS
+
+The AWS STS service will be used to generate temporary credential the user can use to upload a file to S3. The backend
+server will use the STS [assume role](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html)
+functionality to generate a token with temporary access to upload to S3. According to AWS:
+
+> The resulting session's permissions are the intersection of the role's identity-based policies and the session policies.
+
+The IAM policies bellow will be combined to limit the access of the token generate to only allow the bearer to upload the
+file in the session policy resource. The session policy will be generated at the time of the request, and will use the
+asset_uuid generated by the server and the data set_uuid provided in the request.
+
+#### Assumed Role IAM Policy
+
+```json
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "s3:ListMultipartUploadParts",
+        "s3:PutObject",
+        "s3:AbortMultipartUpload"
+      ],
+      "Resource": [
+        "arn:aws:s3:::upload-bucket/"
+      ]
+    }
+    ]
+}
+```
+
+#### IAM Session Policy
+
+```json
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "s3:ListMultipartUploadParts",
+        "s3:PutObject",
+        "s3:AbortMultipartUpload"
+      ],
+      "Resource": [
+        "arn:aws:s3:::upload-bucket/{data set_uuid}/{asset_uuid}"
+      ]
+    }
+    ]
+}
+```
+
+### EC2
+
+The EC2 instance will be use to download files from a CSP and upload to our Upload bucket.
+### Upload Bucket
+
+The upload bucket is a separate bucket or separate bucket location within our main bucket. It is used to stage files
+before moving to the main bucket.
+
+S3 event can be use to trigger additional processing such as validation, once an upload has completed.
+
+### Cloud to Cloud Upload Flow
+
+Cloud to cloud uploads will allow a user to provide a sharable link from their CSP to the file they wish to upload,
+which DP will used to upload the file to our S3 bucket. Now to walk though the cloud to cloud flow, see figure 1.
+
+![Cloud to Cloud](https://app.lucidchart.com/publicSegments/view/e1e72143-c108-4c7f-bf6c-053d98256c1f/image.png)
+**Figure 2:** Architecture of a user uploading from their cloud to our cloud.
+
+#### 1. Get Share Link
+
```

Comment:
During the UX review of the collections flow, we discussed the potential for publishers to select **multiple** datasets (but not directories) for parallel upload and processing.

Reply:
What this tells me as a publisher is I can go to say drop box, select the files I want to upload, generate a single link, and share that with Data Portal. Data Portal will then upload all of those files from that single link. If that is the scenario you are envisioning then that would be the same as uploading a directory. When you generate a shared link to to multiple files in both drop box or google drive, the files organized into a folder. I think we should crawl before we can walk, and require one shared link per file. There are a hand full of non trivial problem that will need to solved before we can support a directory of files, or a link that shares multiple files. 

With some frontend work we could support parallel uploads of local files.